### PR TITLE
[FIX] Email Body retrieval logic in IMAP Listener

### PR DESCRIPTION
**Pull request by ayeshLK**
Fixes [#2777](https://github.com/ballerina-platform/ballerina-standard-library/issues/2777)

## Examples
N/A

## Checklist
- [X] Linked to an issue
- [X] Updated the changelog
- [X] Added tests
- [ ] ~Updated the spec~

### CODE REVIEW

File: native/src/main/java/io/ballerina/stdlib/email/util/EmailAccessUtil.java [hunk 1]
```diff
@@ -351,9 +355,22 @@ private static Object getJsonContent(String messageContent) {
...
+        String contentType = message.getContentType();
+        if (contentType != null && CommonUtil.isTextBased(contentType.toLowerCase(Locale.getDefault()))) {
             if (message.getContent() != null) {
-                messageBody = message.getContent().toString();
+                Object content = message.getContent();
+                if (content instanceof InputStream) {
+                    ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+                    int nRead;
+                    byte[] data = new byte[1024];
+                    while ((nRead = ((InputStream) content).read(data, 0, data.length)) != -1) {
```
Comment (chamil321) [inline native/src/main/java/io/ballerina/stdlib/email/util/EmailAccessUtil.java:366]: Where do we close this InputStream?

Comment (chamil321) [reply native/src/main/java/io/ballerina/stdlib/email/util/EmailAccessUtil.java:366]: Try using `try with resource` feature

Comment (ayeshLK) [reply native/src/main/java/io/ballerina/stdlib/email/util/EmailAccessUtil.java]: Fixed with following commits:
[d3847a2ce4fc78cdea9fd2f31983703007bf25ea](https://github.com/ballerina-platform/module-ballerina-email/pull/468/commits/d3847a2ce4fc78cdea9fd2f31983703007bf25ea)  
[dc04c69ab95635a8f0bdcf0a9745e13827eec37e](https://github.com/ballerina-platform/module-ballerina-email/pull/468/commits/dc04c69ab95635a8f0bdcf0a9745e13827eec37e)

File: native/src/main/java/io/ballerina/stdlib/email/util/EmailAccessUtil.java [hunk 2]
```diff
@@ -350,22 +354,36 @@ private static Object getJsonContent(String messageContent) {
...
-        if (message.getContentType() != null && CommonUtil.isTextBased(message.getContentType())) {
-            if (message.getContent() != null) {
-                messageBody = message.getContent().toString();
+        String contentType = message.getContentType();
+        if (contentType != null && CommonUtil.isTextBased(contentType.toLowerCase(Locale.getDefault()))) {
+            Object content = message.getContent();
+            if (content == null) {
+                return "";
+            }
+            if (message.getContent() instanceof InputStream) {
```
Comment (chamil321) [inline native/src/main/java/io/ballerina/stdlib/email/util/EmailAccessUtil.java:363]:
```suggestion
            if (content instanceof InputStream) {
```

File: native/src/main/java/io/ballerina/stdlib/email/util/EmailAccessUtil.java [hunk 3]
```diff
@@ -350,22 +354,36 @@ private static Object getJsonContent(String messageContent) {
...
+                ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+                int nRead;
+                byte[] data = new byte[1024];
+                while ((nRead = ((InputStream) content).read(data, 0, data.length)) != -1) {
+                    buffer.write(data, 0, nRead);
+                }
+                buffer.flush();
+                ((InputStream) content).close();
```
Comment (chamil321) [inline native/src/main/java/io/ballerina/stdlib/email/util/EmailAccessUtil.java:371]: Avoid multiple casting by using a single new variable

**Comment by shafreenAnfar** [comment]
@ayeshLK and I had a chat and decided to update the testcase names. The fix is simply about making MIME type case insensitive. The testcases are simply there to validate this.

File: native/src/main/java/io/ballerina/stdlib/email/util/EmailAccessUtil.java (hunk 2 above)
Comment (ayeshLK) [reply native/src/main/java/io/ballerina/stdlib/email/util/EmailAccessUtil.java]: Fixed with [aaa1f3095e264cef55954ceb00369e856cf142f7](https://github.com/ballerina-platform/module-ballerina-email/pull/468/commits/aaa1f3095e264cef55954ceb00369e856cf142f7)

File: native/src/main/java/io/ballerina/stdlib/email/util/EmailAccessUtil.java (hunk 3 above)
Comment (ayeshLK) [reply native/src/main/java/io/ballerina/stdlib/email/util/EmailAccessUtil.java]: Fixed with [aaa1f3095e264cef55954ceb00369e856cf142f7](https://github.com/ballerina-platform/module-ballerina-email/pull/468/commits/aaa1f3095e264cef55954ceb00369e856cf142f7)

File: native/src/main/java/io/ballerina/stdlib/email/util/EmailAccessUtil.java [hunk 4]
```diff
@@ -350,22 +354,37 @@ private static Object getJsonContent(String messageContent) {
...
+            Object content = message.getContent();
+            if (content == null) {
+                return "";
+            }
+            if (content instanceof InputStream) {
+                InputStream contentStream = (InputStream) content;
+                ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+                int nRead;
+                byte[] data = new byte[1024];
+                while ((nRead = (contentStream).read(data, 0, data.length)) != -1) {
```
Comment (chamil321) [inline native/src/main/java/io/ballerina/stdlib/email/util/EmailAccessUtil.java:368]:
```suggestion
                while ((nRead = contentStream.read(data, 0, data.length)) != -1) {
```

File: native/src/main/java/io/ballerina/stdlib/email/util/EmailAccessUtil.java [hunk 5]
```diff
@@ -350,22 +354,37 @@ private static Object getJsonContent(String messageContent) {
...
+                    buffer.write(data, 0, nRead);
+                }
+                buffer.flush();
+                (contentStream).close();
```
Comment (chamil321) [inline native/src/main/java/io/ballerina/stdlib/email/util/EmailAccessUtil.java:372]:
```suggestion
                contentStream.close();
```

**Review by chamil321 — APPROVED** [review]

**Pull request merged**